msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java

```diff
@@ -531,57 +525,143 @@ public void skipValue() throws IOException {
      */
     private static MessageTypeException unexpected(String expected, byte b)
             throws MessageTypeException {
-        ValueType type = ValueType.valueOf(b);
-        return new MessageTypeException(String.format("Expected %s, but got %s (%02x)", expected, type.toTypeName(), b));
+        MessageFormat format = MessageFormat.valueOf(b);
+        String typeName;
+        if (format == MessageFormat.NEVER_USED) {
+            typeName = "NeverUsed";
+        } else {
+            String name = format.getValueType().name();
+            typeName = name.substring(0, 1) + name.substring(1).toLowerCase();
+        }
+        return new MessageTypeException(String.format("Expected %s, but got %s (%02x)", expected, typeName, b));
     }
 
-    public MessageFormat unpackValue(ValueHolder holder) throws IOException {
+    public ImmutableValue unpackValue() throws IOException {
         MessageFormat mf = getNextFormat();
         switch(mf.getValueType()) {
             case NIL:
                 unpackNil();
-                holder.setNil();
-                break;
+                return ValueFactory.newNilValue();
             case BOOLEAN:
-                holder.setBoolean(unpackBoolean());
-                break;
-            case INTEGER: {
-                unpackInteger(holder.getIntegerHolder());
-                holder.setToInteger();
-                break;
+                return ValueFactory.newBooleanValue(unpackBoolean());
+            case INTEGER:
+                switch (mf) {
+                    case UINT64:
+                        return ValueFactory.newIntegerValue(unpackBigInteger());
+                    default:
+                        return ValueFactory.newIntegerValue(unpackLong());
+                }
+            case FLOAT:
+                return ValueFactory.newFloatValue(unpackDouble());
```

Comment:
I remembered why FloatHolder was necessary. When reading a float value as double, it will add some fractions. For example: 
```
scala> 0.1341f.toDouble
res5: Double = 0.13410000503063202
```
If user wants to print this float value as String, `0.13410000503063202` will be shown. This is unexpected behavior for the user.

Another example: `0.1f` becomes '0.10000000149011612` if we use double as `0.1f`s internal representation.

I'll fix this in another pull request.